Response File support

bschommer opened the pull request:
This pull request adds response file support to avoid problems with command line limits under Windows. A response file allows it to pass command line arguments per file. Options in the file are separated by whitespace and inserted in place.

The quoting conventions for options follows that of the gnu-tools: Options containing whitespace must be surrounded by either double or single quotes.

Additionally responsefiles are also used to pass options to the assembler, linker and preprocessor.

xavierleroy commented:
I agree we need to support response files for the reasons you give.  Thanks for initiating this effort.

There are a number of delicate points to address:
* Diab 5 tools use the command-line syntax `-@responsefile` while Microsoft tools and GNU tools under Win32 use `@responsefile`.  This must be taken into account when ccomp passes arguments via response files.  Probably this should be configurable via compcert.ini.
* For the various tools called from CompCert, the syntax of response files (quoting conventions, etc) needs to be fully understood.  
  * For Microsoft tools, https://msdn.microsoft.com/en-us/library/3te4xt0y.aspx suggests that response files are parsed exactly like command-line arguments, and the parsing conventions for the latter are here: https://msdn.microsoft.com/en-us/library/a1y7w461.aspx .  Note that response files cannot be nested.
  * For GNU tools, it seems like singe and double quotes are recognized, and backslash systematically escapes the next character (in any context).  No mention of nested response files.
  * For Diab 5 tools, no info is given in the manual.
* As a consequence of the above, `Filename.quote` may not produce syntax that the tool can parse.  E.g. on Win32, the output of `Filename.quote` is OK for the Microsoft tools but not for the GNU tools.
* We also need to agree on a syntax for the response files given to CompCert.  The syntax implemented in this PR (based on an earlier proposal of mine) is Unix-shell-like with the minimal modifications to ensure that the output of `Filename.quote` can be read back both on Unix and Windows.  However, under Win32 it is not compatible with the Microsoft parsing conventions (e.g. `\\` outside a quote is parsed as a single backslash, while it would denote two backslashes for a Microsoft tool).  It is not compatible with the GNU tools either (e.g. no backslash escapes between single quotes).

In a nutshell: it's complicated!  But let's try to find some answers nonetheless.

xavierleroy commented on driver/Assembler.ml:
```diff
@@ -18,12 +18,17 @@ open Driveraux
     List.rev !assembler_options;
     [ifile]
   ] in
+  let opts = responsefile opts (fun a -> if gnu_system then ["@"^a] else ["@"^a]) in
```
Did you mean `"-@"^a` in the `else` branch?  At any rate, it might be better to take the appropriate prefix `@` or `-@` from an entry in `compcert.ini`

xavierleroy replied:
Also: I wonder if this logic (of using a response file or not) could be factored out by putting it in `Driveraux.command` or maybe a new function `Driveraux.command_respfile` on top of `Driveraux.command`.

xavierleroy commented on driver/Driveraux.ml:
```diff
@@ -14,12 +14,33 @@
 
+let responsefile args resp_arg =
+  try
+    if Sys.win32 && (String.length (String.concat "" args) > 7000) then
```
Could do `List.fold_left (fun len arg -> len + String.length arg + 1) 0 args > 7000` here, to avoid constructing the concatenated string.  Also, one space should be counted between every argument, I think.  Finally, I vaguely remember that Cygwin has its own limitations on the command-line length, not as strict as Win32's, but not as big as Linux or MacOS.

xavierleroy commented on driver/Driveraux.ml:
```diff
@@ -14,12 +14,33 @@
+      let file,oc = Filename.open_temp_file "compcert" "" in
+      let whitespace = Str.regexp "[ \t\r\n]" in
+      let quote arg =
+        if Str.string_match whitespace arg 0 then
```
Depending on the syntax for responsefiles implemented by the tool, other characters may need quoting.  E.g. backslashes with the GNU tool syntax.

xavierleroy commented on driver/Frontend.ml:
```diff
@@ -24,15 +24,20 @@ open Printf
     List.rev !prepro_options;
     [ifile]
   ] in
+  let opts = responsefile opts (fun a -> if gnu_system then ["@"^a] else ["@"^a]) in
```
Same comment as above.

xavierleroy commented on driver/Linker.ml:
```diff
@@ -19,14 +19,19 @@ open Driveraux
+    then ["-L" ^ !stdlib_path; "-lcompcert"]
+    else [])
   ] in
+  let opts = responsefile opts (fun a -> if gnu_system then ["@"^a] else ["-@"^a]) in
```
Same comment as above.

xavierleroy commented on driver/Driveraux.ml:
```diff
@@ -14,12 +14,33 @@
+    else
+      args
+  with Sys_error _ ->
+    args
```
When is the temporary response file removed?

bschommer commented:
The responsefile functionality is now merged into command and also the temporary file is removed after the command returns.

Concerning quoting: My suggestion would be to adopt the conventions of the gnu toolchain since most build systems will have support for this. I will further investigate about the exact conventions.

xavierleroy commented:
I agree that for generating @files and passing them to the linker, etc, we must support the GNU syntax, because most of the time it's a GNU toolchain we're using.  But I'm afraid we have to support the Diab syntax as well, which mean reverse-engineering it first...

For parsing @files given to CompCert, we could use the GNU syntax also.  My only reservation is that this syntax is not friendly towards Windows file names (backslash needs escaping, always).

xavierleroy commented on driver/Commandline.ml:
```diff
@@ -99,4 +100,9 @@ let parse_array spec argv first last =
+  try
+    let argv = expandargv Sys.argv in
+    parse_array spec argv 1 (Array.length argv - 1)
+  with Arg.Bad s ->
```
Can the `Arg.Bad` exception be raised here?  I don't think so.  Perhaps you meant `Responsefile.Error`.

xavierleroy commented on driver/Driveraux.ml:
```diff
@@ -14,13 +14,20 @@
+(* Is this a gnu based tool chain *)
+let gnu_system = Configuration.system <> "diab"
+
+(* Sage removale of files *)
```
Two typos in this comment.

xavierleroy commented on driver/Driveraux.ml:
```diff
@@ -51,12 +58,55 @@ let command ?stdout args =
     -1
 
+(* This function reimplements quoting of writeargv from libiberty *)
+let gnu_quote arg =
```
I'd be tempted to move `gnu_quote` and `diab_quote` to the new lib/Responsefile module.  It makes sense to have parsing and unparsing in the same module.

xavierleroy commented on driver/Driveraux.ml:
```diff
@@ -51,12 +58,55 @@ let command ?stdout args =
+    | [] -> assert false (* Should never happen *) in
+    List.iter (fun a -> Printf.fprintf oc "%s " (quote a)) args;
+    close_out oc;
+    let arg = if gnu_system then "@"^file else "-@"^file in
```
It is not consistent to hardwire the choice between `@` and `-@` in the Caml code, while leaving the choice of response file syntax to the `compcert.ini` file.  I mean, either both parameters (response file syntax and prefix to put before the response file name) should be entries of `compcert.ini`, or none.

xavierleroy commented on driver/Linker.ml:
```diff
@@ -24,8 +24,8 @@ let linker exe_name files =
     (if Configuration.has_runtime_lib
-     then ["-L" ^ !stdlib_path; "-lcompcert"]
-     else [])
+    then ["-L" ^ !stdlib_path; "-lcompcert"]
```
It's good to avoid whitespace-only changes.  Plus, the previous indentation is better than the new one.

xavierleroy commented on lib/Responsefile.mll:
```diff
@@ -0,0 +1,97 @@
+(* *********************************************************************)
+
+
+(* Parsing response files with various quoting styles *)
```
This comment no longer matches the contents of the file...  Better remove it.

bschommer commented:
I fixed the things mentioned in the comments and moved the quoting functionality into Responsefile.ml.
The responsfile style now also influences whether `@` or `-@` is used.

xavierleroy commented:
Thanks for the updates.  Just merged.  

Tiny comment: you probably no longer need `let gnu_system = ...` in Driveraux.ml.
